File: MonoInterface/TypeMapChecks/PhysicalEntityFlags.cpp
#include "stdafx.h"

#include "CheckingBasics.h"

TYPE_MIRROR enum _phentity_flags
{
	// PE_PARTICLE-specific flags
	particle_single_contact_check = 0x01, // full stop after first contact
	particle_constant_orientation_check = 0x02, // forces constant orientation
	particle_no_roll_check = 0x04, // 'sliding' mode; entity's 'normal' vector axis will be alinged with the ground normal
	particle_no_path_alignment_check = 0x08, // unless set, entity's y axis will be aligned along the movement trajectory
	particle_no_spin_check = 0x10, // disables spinning while flying
	particle_no_self_collisions_check = 0x100, // disables collisions with other particles
	particle_no_impulse_check = 0x200, // particle will not add hit impulse (expecting that some other system will) 

	// PE_LIVING-specific flags
	lef_push_objects_check = 0x01,
	lef_push_players_check = 0x02,	// push objects and players during contacts
	lef_snap_velocities_check = 0x04,	// quantizes velocities after each step (was ised in MP for precise deterministic sync)
	lef_loosen_stuck_checks_check = 0x08, // don't do additional intersection checks after each step (recommended for NPCs to improve performance)
	lef_report_sliding_contacts_check = 0x10,	// unless set, 'grazing' contacts are not reported 

	// PE_ROPE-specific flags
	rope_findiff_attached_vel_check = 0x01, // approximate velocity of the parent object as v = (pos1-pos0)/time_interval
	rope_no_solver_check = 0x02, // no velocity solver; will rely on stiffness (if set) and positional length enforcement
	rope_ignore_attachments_check = 0x4, // no collisions with objects the rope is attached to
	rope_target_vtx_rel0_check = 0x08,
	rope_target_vtx_rel1_check = 0x10, // whether target vertices are set in the parent entity's frame
	rope_subdivide_segs_check = 0x100, // turns on 'dynamic subdivision' mode (only in this mode contacts in a strained state are handled correctly)
	rope_no_tears_check = 0x200, // rope will not tear when it reaches its force limit, but stretch
	rope_collides_check = 0x200000, // rope will collide with objects other than the terrain
	rope_collides_with_terrain_check = 0x400000, // rope will collide with the terrain
	rope_collides_with_attachment_check = 0x80, // rope will collide with the objects it's attached to even if the other collision flags are not set
	rope_no_stiffness_when_colliding_check = 0x10000000, // rope will use stiffness 0 if it has contacts

	// PE_SOFT-specific flags
	se_skip_longest_edges_check = 0x01,	// the longest edge in each triangle with not participate in the solver
	se_rigid_core_check = 0x02, // soft body will have an additional rigid body core

	// PE_RIGID-specific flags (note that PE_ARTICULATED and PE_WHEELEDVEHICLE are derived from it)
	ref_use_simple_solver_check = 0x01,	// use penalty-based solver (obsolete)
	ref_no_splashes_check = 0x04, // will not generate EventPhysCollisions when contacting water
	ref_checksum_received_check = 0x04,
	ref_checksum_outofsync_check = 0x08, // obsolete
	ref_small_and_fast_check = 0x100, // entity will trace rays against alive characters; set internally unless overriden

	// PE_ARTICULATED-specific flags
	aef_recorded_physics_check = 0x02, // specifies a an entity that contains pre-baked physics simulation

	// PE_WHEELEDVEHICLE-specific flags
	wwef_fake_inner_wheels_check = 0x08, // exclude wheels between the first and the last one from the solver
	// (only wheels with non-0 suspension are considered)

	// general flags
	pef_parts_traceable_check = 0x10,	// each entity part will be registered separately in the entity grid
	pef_disabled_check = 0x20, // entity will not be simulated
	pef_never_break_check = 0x40, // entity will not break or deform other objects
	pef_deforming_check = 0x80, // entity undergoes a dynamic breaking/deforming
	pef_pushable_by_players_check = 0x200, // entity can be pushed by playerd	
	pef_traceable_check = 0x400,
	particle_traceable_check = 0x400,
	rope_traceable_check = 0x400, // entity is registered in the entity grid
	pef_update_check = 0x800, // only entities with this flag are updated if ent_flagged_only is used in TimeStep()
	pef_monitor_state_changes_check = 0x1000, // generate immediate events for simulation class changed (typically rigid bodies falling asleep)
	pef_monitor_collisions_check = 0x2000, // generate immediate events for collisions
	pef_monitor_env_changes_check = 0x4000,	// generate immediate events when something breaks nearby
	pef_never_affect_triggers_check = 0x8000,	// don't generate events when moving through triggers
	pef_invisible_check = 0x10000, // will apply certain optimizations for invisible entities
	pef_ignore_ocean_check = 0x20000, // entity will ignore global water area
	pef_fixed_damping_check = 0x40000,	// entity will force its damping onto the entire group
	pef_monitor_poststep_check = 0x80000, // entity will generate immediate post step events
	pef_always_notify_on_deletion_check = 0x100000,	// when deleted, entity will awake objects around it even if it's not referenced (has refcount 0)
	pef_override_impulse_scale_check = 0x200000, // entity will ignore breakImpulseScale in PhysVars
	pef_players_can_break_check = 0x400000, // playes can break the entiy by bumping into it
	pef_cannot_squash_players_check = 0x10000000,	// entity will never trigger 'squashed' state when colliding with players
	pef_ignore_areas_check = 0x800000, // entity will ignore phys areas (gravity and water)
	pef_log_state_changes_check = 0x1000000, // entity will log simulation class change events
	pef_log_collisions_check = 0x2000000, // entity will log collision events
	pef_log_env_changes_check = 0x4000000, // entity will log EventPhysEnvChange when something breaks nearby
	pef_log_poststep_check = 0x8000000, // entity will log EventPhysPostStep events
};

#define CHECK_ENUM(x) static_assert (_phentity_flags::x ## _check == phentity_flags::x, "phentity_flags enumeration has been changed.")

inline void Check()
{
	CHECK_ENUM(particle_single_contact);
	CHECK_ENUM(particle_constant_orientation);
	CHECK_ENUM(particle_no_roll);
	CHECK_ENUM(particle_no_path_alignment);
	CHECK_ENUM(particle_no_spin);
	CHECK_ENUM(particle_no_self_collisions);
	CHECK_ENUM(particle_no_impulse);

	CHECK_ENUM(lef_push_objects);
	CHECK_ENUM(lef_push_players);
	CHECK_ENUM(lef_snap_velocities);
	CHECK_ENUM(lef_loosen_stuck_checks);
	CHECK_ENUM(lef_report_sliding_contacts);

	CHECK_ENUM(rope_findiff_attached_vel);
	CHECK_ENUM(rope_no_solver);
	CHECK_ENUM(rope_ignore_attachments);
	CHECK_ENUM(rope_target_vtx_rel0);
	CHECK_ENUM(rope_target_vtx_rel1);
	CHECK_ENUM(rope_subdivide_segs);
	CHECK_ENUM(rope_no_tears);
	CHECK_ENUM(rope_collides);
	CHECK_ENUM(rope_collides_with_terrain);
	CHECK_ENUM(rope_collides_with_attachment);
	CHECK_ENUM(rope_no_stiffness_when_colliding);

	CHECK_ENUM(se_skip_longest_edges);
	CHECK_ENUM(se_rigid_core);

	CHECK_ENUM(ref_use_simple_solver);
	CHECK_ENUM(ref_no_splashes);
	CHECK_ENUM(ref_checksum_received);
	CHECK_ENUM(ref_checksum_outofsync);
	CHECK_ENUM(ref_small_and_fast);

	CHECK_ENUM(aef_recorded_physics);

	CHECK_ENUM(wwef_fake_inner_wheels);

	CHECK_ENUM(pef_parts_traceable);
	CHECK_ENUM(pef_disabled);
	CHECK_ENUM(pef_never_break);
	CHECK_ENUM(pef_deforming);
	CHECK_ENUM(pef_pushable_by_players);
	CHECK_ENUM(pef_traceable);
	CHECK_ENUM(particle_traceable);
	CHECK_ENUM(rope_traceable);
	CHECK_ENUM(pef_update);
	CHECK_ENUM(pef_monitor_state_changes);
	CHECK_ENUM(pef_monitor_collisions);
	CHECK_ENUM(pef_monitor_env_changes);
	CHECK_ENUM(pef_never_affect_triggers);
	CHECK_ENUM(pef_invisible);
	CHECK_ENUM(pef_ignore_ocean);
	CHECK_ENUM(pef_fixed_damping);
	CHECK_ENUM(pef_monitor_poststep);
	CHECK_ENUM(pef_always_notify_on_deletion);
	CHECK_ENUM(pef_override_impulse_scale);
	CHECK_ENUM(pef_players_can_break);
	CHECK_ENUM(pef_cannot_squash_players);
	CHECK_ENUM(pef_ignore_areas);
	CHECK_ENUM(pef_log_state_changes);
	CHECK_ENUM(pef_log_collisions);
	CHECK_ENUM(pef_log_env_changes);
	CHECK_ENUM(pef_log_poststep);
}

File: Cryambly/Engine/Physics/PhysicalEntityFlags.cs
using System;

namespace CryCil.Engine.Physics
{
	/// <summary>
	/// Enumeration of flags that can be set for physical entities.
	/// </summary>
	[Flags]
	public enum PhysicalEntityFlags
	{
		/// <summary>
		/// When set, specifies that this physical entity will lose all kinetic energy when makes first
		/// contact with another one.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Particle"/>.</remarks>
		ParticleSingleContact = 0x01,
		/// <summary>
		/// When set, specifies that this physical entity doesn't change its orientation via physical
		/// interactions.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Particle"/>.</remarks>
		ParticleConstantOrientation = 0x02,
		/// <summary>
		/// When set, specifies that this physical entity is moving in 'sliding' mode: it's normal vector
		/// is aligned with a ground normal.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Particle"/>.</remarks>
		ParticleNoRoll = 0x04,
		/// <summary>
		/// When set, specifies that this physical entity's forward axis doesn't need to be aligned with
		/// the movement trajectory.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Particle"/>.</remarks>
		ParticleNoPathAlignment = 0x08,
		/// <summary>
		/// When set, specifies that this physical entity won't change its orientation during flight.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Particle"/>.</remarks>
		ParticleNoSpin = 0x10,
		/// <summary>
		/// When set, specifies that this physical entity doesn't collide with other particles.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Particle"/>.</remarks>
		ParticleNoSelfCollisions = 0x100,
		/// <summary>
		/// When set, specifies that this physical entity will not add hit impulse to other entities when
		/// it collides with them.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Particle"/>.</remarks>
		ParticleNoImpulse = 0x200,

		/// <summary>
		/// When set, specifies that this physical entity will push other objects away when it collides
		/// with them.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Living"/>.</remarks>
		LivingEntityPushObjects = 0x01,
		/// <summary>
		/// When set, specifies that this physical entity will push players away when it collides with
		/// them.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Living"/>.</remarks>
		LivingEntityPushPlayers = 0x02,
		/// <summary>
		/// When set, specifies that this physical entity has its velocity quantized after each step in
		/// physics simulation (was used in MultiPlayer before for precise deterministic synchronization).
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Living"/>.</remarks>
		LivingEntitySnapVelocities = 0x04,
		/// <summary>
		/// When set, specifies that this physical entity doesn't have additional intersection checks done
		/// for it after each step in physics simulation (recommended for NPCs to improve performance, but
		/// increases likelihood of them getting stuck).
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Living"/>.</remarks>
		LivingEntityLoosenStuckChecks = 0x08,
		/// <summary>
		/// When set, specifies that this physical entity will react to 'grazing' contacts with other
		/// entities.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Living"/>.</remarks>
		LivingEntityReportSlidingContacts = 0x10,

		/// <summary>
		/// When set, specifies that this physical entity approximates velocity of the parent object using
		/// formula: <c>v = (pos1 - pos0) / time_span;</c>
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeApproximateAttachedVelocity = 0x01,
		/// <summary>
		/// When set, specifies that this physical entity doesn't use velocity solver and instead relies on
		/// stiffness (if set) and positional length enforcement.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeNoSolver = 0x02,
		/// <summary>
		/// When set, specifies that this physical entity doesn't collide with objects it's attached to.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeIgnoreAttachments = 0x4,
		/// <summary>
		/// When set, specifies that this physical entity targets a vertex in the first attached entity.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeTargetVertexRelative0 = 0x08,
		/// <summary>
		/// When set, specifies that this physical entity targets a vertex in the second attached entity.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeTargetVertexRelative1 = 0x10,
		/// <summary>
		/// When set, specifies that this physical entity has its segments dynamically subdivided (allows
		/// contacts in strained state to be handled correctly).
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeSubdivideSegments = 0x100,
		/// <summary>
		/// When set, specifies that this physical entity won't tear when it reaches its force limit and
		/// will stretch instead.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeNoTears = 0x200,
		/// <summary>
		/// When set, specifies that this physical entity will collide with objects other then terrain.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeCollides = 0x200000,
		/// <summary>
		/// When set, specifies that this physical entity will collide with terrain.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeCollidesWithTerrain = 0x400000,
		/// <summary>
		/// When set, specifies that this physical entity will collide with the objects it's attached to
		/// even if the other collision flags are not set.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeCollidesWithAttachment = 0x80,
		/// <summary>
		/// When set, specifies that this physical entity uses stiffness value of 0 when processing
		/// contacts.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Rope"/>.</remarks>
		RopeNoStiffnessWhenColliding = 0x10000000,

		/// <summary>
		/// When set, specifies that this physical entity will have the longest edge in each triangle
		/// skipped by the solver (increase performance at the cost of precision).
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Soft"/>.</remarks>
		SoftEntitySkipLongestEdges = 0x01,
		/// <summary>
		/// When set, specifies that this physical entity has an additional rigid body core.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Soft"/>.</remarks>
		SoftEntityRigidCore = 0x02,

		/// <summary>
		/// Deprecated.
		/// </summary>
		/// <remarks>
		/// This flag is specific to <see cref="PhysicalEntityType.Rigid"/>,
		/// <see cref="PhysicalEntityType.Articulated"/> and
		/// <see cref="PhysicalEntityType.WheeledVehicle"/>.
		/// </remarks>
		[Obsolete]
		RigidEntityUseSimpleSolver = 0x01,
		/// <summary>
		/// When set, specifies that this physical entity will not report contacts with water.
		/// </summary>
		/// <remarks>
		/// This flag is specific to <see cref="PhysicalEntityType.Rigid"/>,
		/// <see cref="PhysicalEntityType.Articulated"/> and
		/// <see cref="PhysicalEntityType.WheeledVehicle"/>.
		/// </remarks>
		RigidEntityNoSplashes = 0x04,
		/// <summary>
		/// Deprecated.
		/// </summary>
		/// <remarks>
		/// This flag is specific to <see cref="PhysicalEntityType.Rigid"/>,
		/// <see cref="PhysicalEntityType.Articulated"/> and
		/// <see cref="PhysicalEntityType.WheeledVehicle"/>.
		/// </remarks>
		[Obsolete]
		RigidEntityCheckSumReceived = 0x04,
		/// <summary>
		/// Deprecated.
		/// </summary>
		/// <remarks>
		/// This flag is specific to <see cref="PhysicalEntityType.Rigid"/>,
		/// <see cref="PhysicalEntityType.Articulated"/> and
		/// <see cref="PhysicalEntityType.WheeledVehicle"/>.
		/// </remarks>
		[Obsolete]
		RigidEntityCheckSumOutOfSync = 0x08,
		/// <summary>
		/// When set, specifies that this physical entity will trace rays against alive characters (used
		/// for small and fast objects like bullets to make sure they don't go through any players without
		/// touching them). This flag is set internally when the entity is sufficiently small and fast.
		/// </summary>
		/// <remarks>
		/// This flag is specific to <see cref="PhysicalEntityType.Rigid"/>,
		/// <see cref="PhysicalEntityType.Articulated"/> and
		/// <see cref="PhysicalEntityType.WheeledVehicle"/>.
		/// </remarks>
		RigidEntitySmallAndFast = 0x100,

		/// <summary>
		/// When set, specifies that this physical entity contains pre-calculated physics simulation.
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.Articulated"/>.</remarks>
		ArticulatedEntityRecordedPhysics = 0x02,

		/// <summary>
		/// When set, specifies that this physical entity has only its first and last wheel considered by
		/// the solver. (All wheels with non-0 suspension are always considered).
		/// </summary>
		/// <remarks>This flag is specific to <see cref="PhysicalEntityType.WheeledVehicle"/>.</remarks>
		VehicleFakeInnerWheels = 0x08,

		/// <summary>
		/// When set, specifies that this physical entity will have all of its parts registered separately
		/// in the entity grid.
		/// </summary>
		PartsTraceable = 0x10,
		/// <summary>
		/// When set, specifies that this physical entity is not simulated.
		/// </summary>
		Disabled = 0x20,
		/// <summary>
		/// When set, specifies that this physical entity won't break or deform other objects.
		/// </summary>
		NeverBreak = 0x40,
		/// <summary>
		/// When set, specifies that this physical entity can undergo dynamic breakage/deformation.
		/// </summary>
		Deforming = 0x80,
		/// <summary>
		/// When set, specifies that this physical entity can be pushed by the player.
		/// </summary>
		PushableByPlayers = 0x200,
		/// <summary>
		/// When set, specifies that this physical entity is registered in the entity grid.
		/// </summary>
		Traceable = 0x400,
		/// <summary>
		/// When set, specifies that this physical entity is registered in the entity grid as a particle.
		/// </summary>
		ParticleTraceable = 0x400,
		/// <summary>
		/// When set, specifies that this physical entity is registered in the entity grid as a rope.
		/// </summary>
		RopeTraceable = 0x400,
		/// <summary>
		/// When set, specifies that this physical entity will always be updated in each step of the
		/// simulation.
		/// </summary>
		Update = 0x800,
		/// <summary>
		/// When set, specifies that this physical entity generate immediate events for simulation class
		/// changed (typically rigid bodies falling asleep).
		/// </summary>
		MonitorStateChanges = 0x1000,
		/// <summary>
		/// When set, specifies that this physical entity generate immediate events for collisions.
		/// </summary>
		MonitorCollisions = 0x2000,
		/// <summary>
		/// When set, specifies that this physical entity generate immediate events when something breaks
		/// nearby.
		/// </summary>
		MonitorEnvironmentChanges = 0x4000,
		/// <summary>
		/// When set, specifies that this physical entity doesn't report contacts with triggers (
		/// <see cref="PhysicalEntityType.Area"/> type entities?).
		/// </summary>
		NeverAffectTriggers = 0x8000,
		/// <summary>
		/// When set, specifies that this physical entity is invisible (certain optimizations are applied
		/// to these).
		/// </summary>
		Invisible = 0x10000,
		/// <summary>
		/// When set, specifies that this physical entity will ignore the global water volume.
		/// </summary>
		IgnoreOcean = 0x20000,
		/// <summary>
		/// When set, specifies that this physical entity will force its damping onto the entire group.
		/// </summary>
		FixedDamping = 0x40000,
		/// <summary>
		/// When set, specifies that this physical entity will generate immediate post step events.
		/// </summary>
		MonitorPostStep = 0x80000,
		/// <summary>
		/// When set, specifies that this physical entity will awake objects around it when deleted.
		/// </summary>
		AlwaysNotifyOnDeletion = 0x100000,
		/// <summary>
		/// When set, specifies that this physical entity will ignore
		/// <see cref="PhysicsVariables.BreakImpulseScale"/>.
		/// </summary>
		OverrideImpulseScale = 0x200000,
		/// <summary>
		/// When set, specifies that this physical entity can be broken by player that bump into it.
		/// </summary>
		PlayersCanBreak = 0x400000,
		/// <summary>
		/// When set, specifies that this physical entity cannot trigger 'squashed' state when colliding
		/// with players.
		/// </summary>
		CannotSquashPlayers = 0x10000000,
		/// <summary>
		/// When set, specifies that this physical entity will ignore physical areas (gravity and water).
		/// </summary>
		IgnoreAreas = 0x800000,
		/// <summary>
		/// When set, specifies that this physical entity log simulation class change events.
		/// </summary>
		LogStateChanges = 0x1000000,
		/// <summary>
		/// When set, specifies that this physical entity log collision events.
		/// </summary>
		LogCollisions = 0x2000000,
		/// <summary>
		/// When set, specifies that this physical entity log changes when something breaks nearby
		/// </summary>
		LogEnvironmentChanges = 0x4000000,
		/// <summary>
		/// When set, specifies that this physical entity log post-step-update events
		/// </summary>
		LogPostStep = 0x8000000,
	}
}